### PR TITLE
#9803 - GetVRFInputData throws a vector error when using multiple VRF types

### LINKED ISSUE #9803: GetVRFInputData throws a vector error when using multiple VRF types
Issue overview
--------------

Defect file with one AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl and one AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl:HR throws. Looks like an indexing issue.

https://github.com/NREL/EnergyPlus/blob/281e0c092b0377181ca9f738887479a81a52910d/src/EnergyPlus/HVACVariableRefrigerantFlow.cc#L2950-L2967

https://github.com/NREL/EnergyPlus/blob/281e0c092b0377181ca9f738887479a81a52910d/src/EnergyPlus/InputProcessing/InputProcessor.cc#L947

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): all
 - Version of EnergyPlus (if using an intermediate build, include SHA): develop, 281e0c092b0377181ca9f738887479a81a52910d

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added: https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/9000s/9803
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

 - Fixes #9803
 - Refactor GetVRFInputData to limit scope (eliminate variables declared uninitialized atop the routine)

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by jmarrec** [comment]
Running the Defect File on EnergyPlusDevSupport: before fix it throws the vector error, after it fatals out due to other problems in the file

image

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 1]
```diff
@@ -1421,143 +1422,122 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
 
     static constexpr std::string_view RoutineName("GetVRFInput: "); // include trailing blank space
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1423]: Oh, this is going to be fun!

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 2]
```diff
@@ -1569,51 +1549,66 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
+        thisTUList.HeatingCoilAvailable = false;
+        thisTUList.CoolingCoilAvailSchPtr = -1;
+        thisTUList.HeatingCoilAvailSchPtr = -1;
+
+        for (int TUNum = 1; TUNum <= thisTUList.NumTUInList; ++TUNum) {
+            thisTUList.ZoneTUName(TUNum) = cAlphaArgs(TUNum + 1);
+        }
+    }
+
+    auto checkCurveMinMaxOutput = [&state](int CurveIndex) -> std::array<Real64, 4> {
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1588]: Man, you are really loving this lambda trick.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 3]
```diff
@@ -1421,143 +1422,122 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
+        state.dataHVACVarRefFlow->NumVRFCond_FluidTCtrl_HR =
+            state.dataInputProcessing->inputProcessor->getNumObjectsFound(state, "AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl:HR");
+        if (state.dataHVACVarRefFlow->NumVRFCond_FluidTCtrl_HR > 0) {
+            state.dataInputProcessing->inputProcessor->getObjectDefMaxArgs(
+                state, "AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl:HR", NumParams, NumAlphas, NumNums);
+            MaxAlphas = max(MaxAlphas, NumAlphas);
+            MaxNumbers = max(MaxNumbers, NumNums);
+        }
+
+        state.dataHVACVarRefFlow->NumVRFCond = state.dataHVACVarRefFlow->NumVRFCond_SysCurve + state.dataHVACVarRefFlow->NumVRFCond_FluidTCtrl_HP +
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1491]: Wanna see something cute?

```cpp
enum class VRFCondType { Invalid = -1, SysCurve, FluidCtrl_HP, FluidCtrl_HR, Num };
static constexpr std::array<std::string_view, static_cast<int>> VRFCondTypeNames = { "AirConditioner:VariableRefrigerantFlow", "AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl", "AirConditioner:VariableRefrigerantFlow:FluidTemperatureControl:HR" };

MaxAlphas = MaxNums = 0;
for (VRFCondType vrfCondType : {VRFCondType::SysCurve, VRFCondType::FluidCtrl_HP, VRFCondType::FluidCtrl_HR}) {
  state.dataHVACVarRefFlow->NumVRFCondType[static_cast<int>(vrfCondType)] = getNumObjectsFuond(state, VRFCondTypeNames[static_cast<int>(vrfCondType)]);
  if (state.dataHVACVarRefFlow->NumVRFCondType[static_cast<int>(vrfCondType)] > 0) {
      getObjectDefMaxArgs(state, VRFCondTypeNames[static_cast<int>(vrfCondType), ...);
      MaxAlphas = max(MaxAlphas, NumAlphas);
      MaxNums = max(MaxNums, NumNums);
   }
```

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 4]
```diff
@@ -1569,51 +1549,66 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-        state.dataHVACVarRefFlow->TerminalUnitList(VRFNum).HeatingCoilAvailSchPtr = -1;
-
-        for (TUListNum = 1; TUListNum <= state.dataHVACVarRefFlow->TerminalUnitList(VRFNum).NumTUInList; ++TUListNum) {
-            state.dataHVACVarRefFlow->TerminalUnitList(VRFNum).ZoneTUName(TUListNum) = cAlphaArgs(TUListNum + 1);
-        }
-    }
+        auto &thisTUList = state.dataHVACVarRefFlow->TerminalUnitList(TUListNum);
+        thisTUList.Name = cAlphaArgs(1);
+        thisTUList.NumTUInList = NumAlphas - 1;
+        thisTUList.ZoneTUPtr.allocate(thisTUList.NumTUInList);
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1555]: May be better to turn this into a single array of `struct`.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 5]
```diff
@@ -1629,237 +1624,223 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
         if (lAlphaFieldBlanks(2)) {
-            state.dataHVACVarRefFlow->VRF(VRFNum).SchedPtr = DataGlobalConstants::ScheduleAlwaysOn;
+            thisVrfSys.SchedPtr = DataGlobalConstants::ScheduleAlwaysOn;
         } else {
-            state.dataHVACVarRefFlow->VRF(VRFNum).SchedPtr = GetScheduleIndex(state, cAlphaArgs(2));
-            if (state.dataHVACVarRefFlow->VRF(VRFNum).SchedPtr == 0) {
-                ShowSevereError(state, cCurrentModuleObject + "=\"" + state.dataHVACVarRefFlow->VRF(VRFNum).Name + "\" invalid data");
+            thisVrfSys.SchedPtr = GetScheduleIndex(state, cAlphaArgs(2));
+            if (thisVrfSys.SchedPtr == 0) {
+                ShowSevereError(state, cCurrentModuleObject + "=\"" + thisVrfSys.Name + "\" invalid data");
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: `format`

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 6]
```diff
@@ -1629,237 +1624,223 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
                                       format("...Curve minimum must be >= 0.7, curve min at PLR = {:.2T} is {:.3T}", MinCurvePLR, MinCurveVal));
                     ShowContinueError(state, "...Setting curve minimum to 0.7 and simulation continues.");
-                    SetCurveOutputMinMaxValues(state, state.dataHVACVarRefFlow->VRF(VRFNum).CoolPLFFPLR, ErrorsFound, 0.7, _);
+                    SetCurveOutputMinMaxValues(state, thisVrfSys.CoolPLFFPLR, ErrorsFound, 0.7, _);
                 }
 
                 if (MaxCurveVal > 1.0) {
-                    ShowWarningError(
-                        state, std::string{RoutineName} + cCurrentModuleObject + "=\"" + state.dataHVACVarRefFlow->VRF(VRFNum).Name + "\", invalid");
+                    ShowWarningError(state, std::string{RoutineName} + cCurrentModuleObject + "=\"" + thisVrfSys.Name + "\", invalid");
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: #AllTheFomats

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 7]
```diff
@@ -1868,181 +1849,164 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-                                                 cAlphaFieldNames(18));                             // Field Name
+                                                 thisVrfSys.HeatEIRFTHi, // Curve index
+                                                 {2},                    // Valid dimensions
+                                                 RoutineName,            // Routine name
+                                                 cCurrentModuleObject,   // Object Type
+                                                 thisVrfSys.Name,        // Object Name
+                                                 cAlphaFieldNames(18));  // Field Name
         }
 
         if (UtilityRoutines::SameString(cAlphaArgs(19), "WETBULBTEMPERATURE")) {
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1912]: `enum class` + `getEnumerationValue`

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 8]
```diff
@@ -2051,485 +2015,439 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
                 ShowContinueError(state,
                                   format("...Curve maximum value of X = {:.3T} should be 1 and will result in lower energy use than expected.",
                                          maxEIRfLowPLRXInput));
             }
         }
 
-        state.dataHVACVarRefFlow->VRF(VRFNum).MasterZonePtr = UtilityRoutines::FindItemInList(cAlphaArgs(24), state.dataHeatBal->Zone);
+        thisVrfSys.MasterZonePtr = UtilityRoutines::FindItemInList(cAlphaArgs(24), state.dataHeatBal->Zone);
 
         if (UtilityRoutines::SameString(cAlphaArgs(25), "LoadPriority")) {
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: `getEnumerationValue`

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 9]
```diff
@@ -2051,485 +2015,439 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
+        thisVrfSys.CompressorSizeRatio = rNumericArgs(18);
+        thisVrfSys.MaxOATCCHeater = rNumericArgs(19);
 
         if (!lAlphaFieldBlanks(31)) {
-            if (UtilityRoutines::SameString(cAlphaArgs(31), "ReverseCycle"))
-                state.dataHVACVarRefFlow->VRF(VRFNum).DefrostStrategy = StandardRatings::DefrostStrat::ReverseCycle;
-            if (UtilityRoutines::SameString(cAlphaArgs(31), "Resistive"))
-                state.dataHVACVarRefFlow->VRF(VRFNum).DefrostStrategy = StandardRatings::DefrostStrat::Resistive;
-            if (state.dataHVACVarRefFlow->VRF(VRFNum).DefrostStrategy == StandardRatings::DefrostStrat::Invalid) {
+            if (UtilityRoutines::SameString(cAlphaArgs(31), "ReverseCycle")) thisVrfSys.DefrostStrategy = StandardRatings::DefrostStrat::ReverseCycle;
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: `getEnumerationValue`

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 10]
```diff
@@ -2542,154 +2460,148 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
 
-        state.dataHVACVarRefFlow->VRF(VRFNum).Name = cAlphaArgs(1);
-        state.dataHVACVarRefFlow->VRF(VRFNum).VRFSystemTypeNum = VRF_HeatPump;
-        state.dataHVACVarRefFlow->VRF(VRFNum).VRFAlgorithmTypeNum = AlgorithmType::FluidTCtrl;
-        state.dataHVACVarRefFlow->VRF(VRFNum).FuelType = "Electricity";
-        state.dataHVACVarRefFlow->VRF(VRFNum).FuelTypeNum = DataGlobalConstants::ResourceType::Electricity;
+        int VRFNum = state.dataHVACVarRefFlow->NumVRFCond_SysCurve + thisNum;
+        auto &thisVrfFluidCtrl = state.dataHVACVarRefFlow->VRF(VRFNum);
+        thisVrfFluidCtrl.Name = cAlphaArgs(1);
+        thisVrfFluidCtrl.VRFSystemTypeNum = VRF_HeatPump;
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:2472]: `enum class` and remove `Num` (a `Type` is a `Num` already).

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 11]
```diff
@@ -2542,154 +2460,148 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
+        thisVrfFluidCtrl.VRFAlgorithmTypeNum = AlgorithmType::FluidTCtrl;
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: Remove `Num`.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 12]
```diff
@@ -2542,154 +2460,148 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
+        thisVrfFluidCtrl.FuelType = "Electricity";
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:2474]: Should not need this ultimately. Can always lookup in `ResourceTypeNames` when needed.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 13]
```diff
@@ -2542,154 +2460,148 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
+        thisVrfFluidCtrl.FuelTypeNum = DataGlobalConstants::ResourceType::Electricity;
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:2475]: Remove `Num`.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 14]
```diff
@@ -2542,154 +2460,148 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
             ErrorsFound = true;
         }
 
         // Reference OU SH/SC
-        state.dataHVACVarRefFlow->VRF(VRFNum).SH = rNumericArgs(7);
-        state.dataHVACVarRefFlow->VRF(VRFNum).SC = rNumericArgs(8);
+        thisVrfFluidCtrl.SH = rNumericArgs(7);
+        thisVrfFluidCtrl.SC = rNumericArgs(8);
 
         if (UtilityRoutines::SameString(cAlphaArgs(5), "VariableTemp")) {
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:2548]: Excuse me?

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 15]
```diff
@@ -2737,161 +2646,146 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
 
         // Pipe parameters
-        state.dataHVACVarRefFlow->VRF(VRFNum).RefPipDiaSuc = rNumericArgs(17);
-        state.dataHVACVarRefFlow->VRF(VRFNum).RefPipDiaDis = rNumericArgs(17);
-        state.dataHVACVarRefFlow->VRF(VRFNum).RefPipLen = rNumericArgs(18);
-        state.dataHVACVarRefFlow->VRF(VRFNum).RefPipEquLen = rNumericArgs(19);
-        state.dataHVACVarRefFlow->VRF(VRFNum).RefPipHei = rNumericArgs(20);
-        state.dataHVACVarRefFlow->VRF(VRFNum).RefPipInsThi = rNumericArgs(21);
-        state.dataHVACVarRefFlow->VRF(VRFNum).RefPipInsCon = rNumericArgs(22);
+        thisVrfFluidCtrl.RefPipDiaSuc = rNumericArgs(17);
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:2655]: All words must be shortened to three letters.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 16]
```diff
@@ -2737,161 +2646,146 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
+            thisVrfFluidCtrl.DefrostStrategy = StandardRatings::DefrostStrat::ReverseCycle;
         }
 
         if (!lAlphaFieldBlanks(9)) {
-            if (UtilityRoutines::SameString(cAlphaArgs(9), "Timed"))
-                state.dataHVACVarRefFlow->VRF(VRFNum).DefrostControl = StandardRatings::HPdefrostControl::Timed;
-            if (UtilityRoutines::SameString(cAlphaArgs(9), "OnDemand"))
-                state.dataHVACVarRefFlow->VRF(VRFNum).DefrostControl = StandardRatings::HPdefrostControl::OnDemand;
-            if (state.dataHVACVarRefFlow->VRF(VRFNum).DefrostControl == StandardRatings::HPdefrostControl::Invalid) {
+            if (UtilityRoutines::SameString(cAlphaArgs(9), "Timed")) thisVrfFluidCtrl.DefrostControl = StandardRatings::HPdefrostControl::Timed;
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:2698]: `GetEnumerationValue`

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 17]
```diff
@@ -2737,161 +2646,146 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-                             cCurrentModuleObject + ", \"" + state.dataHVACVarRefFlow->VRF(VRFNum).Name + "\" " + cNumericFieldNames(28) +
+                             cCurrentModuleObject + ", \"" + thisVrfFluidCtrl.Name + "\" " + cNumericFieldNames(28) +
                                  " = 0.0 for defrost strategy = RESISTIVE.");
         }
 
-        state.dataHVACVarRefFlow->VRF(VRFNum).CompMaxDeltaP = rNumericArgs(30);
+        thisVrfFluidCtrl.CompMaxDeltaP = rNumericArgs(30);
 
         //@@ The control type
         std::string ThermostatPriorityType = "LoadPriority"; // cAlphaArgs( 25 )
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:2749]: WTH is this?

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 18]
```diff
@@ -2968,220 +2856,207 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-            state.dataHVACVarRefFlow->VRF(VRFNum).MaxOATHeatRecovery =
-                max(state.dataHVACVarRefFlow->VRF(VRFNum).MaxOATCooling, state.dataHVACVarRefFlow->VRF(VRFNum).MaxOATHeating);
-            ShowContinueError(state,
-                              format("... adjusted {} = {:.2T} C", cNumericFieldNames(8), state.dataHVACVarRefFlow->VRF(VRFNum).MaxOATHeatRecovery));
+            thisVrfFluidCtrlHR.MaxOATHeatRecovery = max(thisVrfFluidCtrlHR.MaxOATCooling, thisVrfFluidCtrlHR.MaxOATHeating);
+            ShowContinueError(state, format("... adjusted {} = {:.2T} C", cNumericFieldNames(8), thisVrfFluidCtrlHR.MaxOATHeatRecovery));
         }
 
         // IU Control Type
         if (UtilityRoutines::SameString(cAlphaArgs(5), "VariableTemp")) {
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:2985]: 😡

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 19]
```diff
@@ -3452,135 +3308,130 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-        if (UtilityRoutines::SameString(cAlphaArgs(6), "BlowThrough")) state.dataHVACVarRefFlow->VRFTU(VRFTUNum).FanPlace = DataHVACGlobals::BlowThru;
-        if (UtilityRoutines::SameString(cAlphaArgs(6), "DrawThrough")) state.dataHVACVarRefFlow->VRFTU(VRFTUNum).FanPlace = DataHVACGlobals::DrawThru;
+        if (UtilityRoutines::SameString(cAlphaArgs(6), "BlowThrough")) thisVrfTU.FanPlace = DataHVACGlobals::BlowThru;
+        if (UtilityRoutines::SameString(cAlphaArgs(6), "DrawThrough")) thisVrfTU.FanPlace = DataHVACGlobals::DrawThru;
 
         if (!lAlphaFieldBlanks(7) && !lAlphaFieldBlanks(8)) {
             // Get fan data
-            FanType = cAlphaArgs(7);
-            FanName = cAlphaArgs(8);
+            std::string FanType = cAlphaArgs(7);
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:3401]: :rage:

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 20]
```diff
@@ -3663,181 +3509,164 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
                             }
 
                         } else {
-                            ShowSevereError(state, cCurrentModuleObject + " \"" + state.dataHVACVarRefFlow->VRFTU(VRFTUNum).Name + "\"");
+                            ShowSevereError(state, cCurrentModuleObject + " \"" + thisVrfTU.Name + "\"");
                             ShowContinueError(state,
-                                              "... when checking " +
-                                                  DataHVACGlobals::cAllCoilTypes(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).DXCoolCoilType_Num) +
-                                                  " \"" + cAlphaArgs(12) + "\"");
+                                              "... when checking " + DataHVACGlobals::cAllCoilTypes(thisVrfTU.DXCoolCoilType_Num) + " \"" +
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:3678]: :rage:

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 21]
```diff
@@ -3938,83 +3749,85 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
         } else {
-            if (state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum > 0) {
-                if (state.dataHVACVarRefFlow->VRF(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum).VRFAlgorithmTypeNum ==
-                    AlgorithmType::FluidTCtrl) {
+            if (thisVrfTU.VRFSysNum > 0) {
+                if (state.dataHVACVarRefFlow->VRF(thisVrfTU.VRFSysNum).VRFAlgorithmTypeNum == AlgorithmType::FluidTCtrl) {
                     // Algorithm Type: VRF model based on physics, applicable for Fluid Temperature Control
 
-                    if (UtilityRoutines::SameString(DataHVACGlobals::cAllCoilTypes(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).DXHeatCoilType_Num),
+                    if (UtilityRoutines::SameString(DataHVACGlobals::cAllCoilTypes(thisVrfTU.DXHeatCoilType_Num),
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:3781]: :rage:

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 22]
```diff
@@ -4023,23 +3836,11 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-                                _,
-                                _,
-                                _,
-                                _,
-                                _,
-                                state.dataHVACVarRefFlow->VRF(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum).HeatingPerformanceOATType);
+                                                 _,
+                                                 state.dataHVACVarRefFlow->VRF(thisVrfTU.VRFSysNum).HeatingPerformanceOATType);
                             // Set defrost controls in child object to trip child object defrost calculations
                             SetDXCoolingCoilData(state,
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:3852]: Can I say something? `SetDXCoolingCoilData` is a ridiculous function. Get rid of this pattern pronto.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 23]
```diff
@@ -4181,38 +3978,35 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
-                                    _,
-                                    _,
-                                    _,
-                                    _,
-                                    _,
-                                    _,
-                                    state.dataHVACVarRefFlow->VRF(state.dataHVACVarRefFlow->VRFTU(VRFTUNum).VRFSysNum).HeatingCapacitySizeRatio);
+                                                     thisVrfTU.HeatingCapacitySizeRatio);
+                            } else if (state.dataHVACVarRefFlow->VRF(thisVrfTU.VRFSysNum).HeatingCapacitySizeRatio > 1.0) {
+                                SetDXCoolingCoilData(state,
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:3993]: There are many crazy things in EnergyPlus, but this is top 5. I want to set a single field, so I need to set up and call a function with 20 parameters. Yes, that makes sense.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc [hunk 24]
```diff
@@ -4503,446 +4265,386 @@ void GetVRFInputData(EnergyPlusData &state, bool &ErrorsFound)
...
             errFlag = false;
-            if (UtilityRoutines::SameString(SuppHeatingCoilType, "Coil:Heating:Water")) {
-                state.dataHVACVarRefFlow->VRFTU(VRFTUNum).SuppHeatCoilType_Num = DataHVACGlobals::Coil_HeatingWater;
-            } else if (UtilityRoutines::SameString(SuppHeatingCoilType, "Coil:Heating:Steam")) {
-                state.dataHVACVarRefFlow->VRFTU(VRFTUNum).SuppHeatCoilType_Num = DataHVACGlobals::Coil_HeatingSteam;
-            } else if (UtilityRoutines::SameString(SuppHeatingCoilType, "Coil:Heating:Fuel") ||
-                       UtilityRoutines::SameString(SuppHeatingCoilType, "Coil:Heating:Electric")) {
-                state.dataHVACVarRefFlow->VRFTU(VRFTUNum).SuppHeatCoilType_Num =
-                    HeatingCoils::GetHeatingCoilTypeNum(state, SuppHeatingCoilType, SuppHeatingCoilName, errFlag);
+            if (UtilityRoutines::SameString(thisVrfTU.SuppHeatCoilType, "Coil:Heating:Water")) {
```
Comment (amirroth) [inline src/EnergyPlus/HVACVariableRefrigerantFlow.cc:4308]: :rage:

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc (hunk 2 above)
Comment (jmarrec) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1588]: It avoids duplicating the code, yet making a full function is too much, so a lambda is a good candidate no?

Comment (amirroth) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc:1588]: I'm just saying.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc (hunk 6 above)
Comment (jmarrec) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: I've started doing this, I'm like 1 hour in, and I haven't left the block that deals with the VRF Sys type (so I need to do FluidCtrl and FluidCtrl_HR which are equally as big). I agree with all your comments, but this is a bug fix, so I'm afraid I won't address them.

I feel like we should be able to (at least partly) automate this, either via scripting or via clang-tidy (similar to https://github.com/mikecrowe/clang-tidy-fmt)

Comment (amirroth) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: I get it. I actually want to establish a different pattern entirely for IDF errors, so maybe this format stuff and that will happen in one pass.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc (hunk 20 above)
Comment (amirroth) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc:3678]: You can ignore `format` related comments. Too many of them.

File: src/EnergyPlus/HVACVariableRefrigerantFlow.cc (hunk 6 above)
Comment (jmarrec) [reply src/EnergyPlus/HVACVariableRefrigerantFlow.cc]: cf 7dbcaa0db9cde14a3b4a86f9acac005f8a9b92c3 . It's not impossible that I have done a couple of misteps, I am way too tired to properly check that I've done all these format replacements correctly.

**Comment by Myoldmopar** [comment]
The index fix is fine.  Why were the CI jobs cancelled?

**Comment by jmarrec** [comment]
Ubuntu 18.04 is on a brown-out preceeding the final removal.

**Comment by Myoldmopar** [comment]
This is all good.  CI is happy, with the build warning from a different branch, and the GPU diffs being the only regression.  Merging this.

**Pull request merged**